File: shared/src/commonMain/kotlin/org/nihongo/mochi/ui/games/crossword/CrosswordGameScreen.kt
package org.nihongo.mochi.ui.games.crossword

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectTransformGestures
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.jetbrains.compose.resources.stringResource
import org.nihongo.mochi.presentation.MochiBackground
import org.nihongo.mochi.shared.generated.resources.*
import org.nihongo.mochi.ui.components.GameHUD
import org.nihongo.mochi.ui.components.GameResultOverlay

@Composable
fun CrosswordGameScreen(
    viewModel: CrosswordViewModel,
    onBackClick: () -> Unit
) {
    val cells by viewModel.cells.collectAsState()
    val placedWords by viewModel.placedWords.collectAsState()
    val selectedCell by viewModel.selectedCell.collectAsState()
    val keyboardKeys by viewModel.keyboardKeys.collectAsState()
    val gameTimeSeconds by viewModel.gameTimeSeconds.collectAsState()
    val selectedHintType by viewModel.selectedHintType.collectAsState()
    val selectedMode by viewModel.selectedMode.collectAsState()
    val isFinished by viewModel.isFinished.collectAsState()
    
    val cellsMap = remember(cells) {
        cells.associateBy { it.r to it.c }
    }

    var scale by remember { mutableStateOf(1f) }
    var offset by remember { mutableStateOf(Offset.Zero) }

    val gridSize = 16
    val cellSize = 40.dp
    val gridDimension = cellSize * gridSize

    val currentWord = remember(selectedCell, placedWords) {
        selectedCell?.let { (r, c) ->
            placedWords.find { word ->
                if (word.isHorizontal) {
                    word.row == r && c >= word.col && c < word.col + word.word.length
                } else {
                    word.col == c && r >= word.row && r < word.row + word.word.length
                }
            }
        }
    }

    MochiBackground {
        Box(modifier = Modifier.fillMaxSize()) {
            Column(modifier = Modifier.fillMaxSize()) {
                // Factorized HUD
                GameHUD(
                    primaryStat = stringResource(Res.string.game_crosswords_title) to "",
                    timerFlow = viewModel.gameTimeSeconds,
                    initialTimerValue = gameTimeSeconds
                )

                Box(
                    modifier = Modifier
                        .weight(1f)
                        .fillMaxWidth()
                        .pointerInput(Unit) {
                            detectTransformGestures { _, pan, zoom, _ ->
                                scale *= zoom
                                scale = scale.coerceIn(0.5f, 3f)
                                offset += pan
                            }
                        }
                        .graphicsLayer(
                            scaleX = scale,
                            scaleY = scale,
                            translationX = offset.x,
                            translationY = offset.y
                        ),
                    contentAlignment = Alignment.Center
                ) {
                    Box(modifier = Modifier.wrapContentSize(unbounded = true)) {
                        Column(modifier = Modifier.requiredSize(gridDimension)) {
                            for (r in 0 until gridSize) {
                                Row(modifier = Modifier.wrapContentWidth(unbounded = true)) {
                                    for (c in 0 until gridSize) {
                                        val cell = cellsMap[r to c]
                                        CrosswordCellView(
                                            cell = cell,
                                            isSelected = selectedCell?.first == r && selectedCell?.second == c,
                                            onClick = { 
                                                if (cell != null && !cell.isBlack) {
                                                    viewModel.onCellSelected(r, c)
                                                }
                                            }
                                        )
                                    }
                                }
                            }
                        }
                    }
                }

                // Bottom Panel (Clues and Keyboard)
                Column(
                    modifier = Modifier.fillMaxWidth()
                ) {
                    if (currentWord != null) {
                        Surface(
                            modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 8.dp),
                            shape = MaterialTheme.shapes.medium,
                            color = MaterialTheme.colorScheme.secondaryContainer,
                            shadowElevation = 4.dp
                        ) {
                            Column(modifier = Modifier.padding(12.dp).fillMaxWidth()) {
                                if (selectedMode == CrosswordMode.KANJIS) {
                                    Text(
                                        text = currentWord.phonetics,
                                        modifier = Modifier.fillMaxWidth(),
                                        textAlign = TextAlign.Center,
                                        style = MaterialTheme.typography.headlineSmall,
                                        fontWeight = FontWeight.Bold,
                                        color = MaterialTheme.colorScheme.primary
                                    )
                                    if (currentWord.meaning.isNotEmpty()) {
                                        Text(
                                            text = currentWord.meaning,
                                            modifier = Modifier.fillMaxWidth().padding(top = 4.dp),
                                            textAlign = TextAlign.Center,
                                            style = MaterialTheme.typography.bodyMedium,
                                            color = MaterialTheme.colorScheme.onSecondaryContainer
                                        )
                                    }
                                } else {
                                    val clueText = if (selectedHintType == CrosswordHintType.KANJI) {
                                        currentWord.kanji
                                    } else {
                                        currentWord.meaning
                                    }
                                    Text(
                                        text = clueText,
                                        modifier = Modifier.fillMaxWidth(),
                                        textAlign = TextAlign.Center,
                                        style = MaterialTheme.typography.bodyLarge,
                                        fontWeight = FontWeight.Bold,
                                        maxLines = 2,
                                        overflow = TextOverflow.Ellipsis
                                    )
                                }
                            }
                        }
                    }

                    if (selectedCell != null && !isFinished) {
                        Surface(
                            modifier = Modifier.fillMaxWidth(),
                            color = MaterialTheme.colorScheme.surface.copy(alpha = 0.95f),
                            shadowElevation = 16.dp,
                            tonalElevation = 8.dp
                        ) {
                            Row(
                                modifier = Modifier.padding(8.dp).height(200.dp),
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                LazyVerticalGrid(
                                    columns = GridCells.Fixed(5),
                                    modifier = Modifier.weight(1f),
                                    contentPadding = PaddingValues(4.dp),
                                    horizontalArrangement = Arrangement.spacedBy(4.dp),
                                    verticalArrangement = Arrangement.spacedBy(4.dp)
                                ) {
                                    items(keyboardKeys) { key ->
                                        KeyButton(text = key, onClick = { viewModel.onKeyTyped(key) })
                                    }
                                }
                                
                                Button(
                                    onClick = { viewModel.onDelete() },
                                    modifier = Modifier.width(60.dp).fillMaxHeight().padding(4.dp),
                                    colors = ButtonDefaults.buttonColors(containerColor = MaterialTheme.colorScheme.errorContainer),
                                    contentPadding = PaddingValues(0.dp)
                                ) {
                                    Text("⌫", color = MaterialTheme.colorScheme.onErrorContainer, fontSize = 24.sp)
                                }
                            }
                        }
                    }
                }
            }
            
            if (isFinished) {
                GameResultOverlay(
                    isVictory = true,
                    stats = listOf(
                        stringResource(Res.string.game_crossword_history_item, placedWords.size, selectedMode.name) to formatGameTimeHUD(gameTimeSeconds)
                    ),
                    onReplayClick = { viewModel.startGame { /* Already in game screen */ } },
                    onMenuClick = onBackClick
                )
            }
        }
    }
}

private fun formatGameTimeHUD(seconds: Int): String {
    val m = seconds / 60
    val s = seconds % 60
    return if (m > 0) "${m}m ${s}s" else "${s}s"
}

@Composable
fun KeyButton(text: String, onClick: () -> Unit) {
    Surface(
        modifier = Modifier.aspectRatio(1f).clickable { onClick() },
        shape = MaterialTheme.shapes.small,
        color = MaterialTheme.colorScheme.primaryContainer,
        tonalElevation = 2.dp
    ) {
        Box(contentAlignment = Alignment.Center) {
            Text(text = text, fontSize = 20.sp, fontWeight = FontWeight.Medium)
        }
    }
}

@Composable
fun CrosswordCellView(
    cell: CrosswordCell?,
    isSelected: Boolean,
    onClick: () -> Unit
) {
    val cellSize = 40.dp
    
    // Conteneur fixe pour garder la grille alignée
    Box(
        modifier = Modifier.size(cellSize),
        contentAlignment = Alignment.Center
    ) {
        if (cell != null && !cell.isBlack) {
            val isInputCorrect = cell.userInput.isNotEmpty() && cell.userInput == cell.solution
            val isInputWrong = cell.userInput.isNotEmpty() && cell.userInput != cell.solution

            val backgroundColor = when {
                cell.isCorrect || isInputCorrect -> Color(0xFFC8E6C9)
                isInputWrong -> Color(0xFFFFCDD2)
                isSelected -> MaterialTheme.colorScheme.primaryContainer
                else -> Color.White
            }

            // Surface gère l'élévation et l'ombre correctement
            Surface(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(1.dp) // Important: laisse l'ombre visible sans chevaucher les voisins
                    .clickable { onClick() },
                shape = RoundedCornerShape(2.dp),
                color = backgroundColor,
                shadowElevation = if (isSelected) 4.dp else 2.dp,
                border = BorderStroke(
                    width = if (isSelected) 1.5.dp else 0.5.dp,
                    color = if (isSelected) MaterialTheme.colorScheme.primary else Color.Gray.copy(alpha = 0.5f)
                )
            ) {
                Box(contentAlignment = Alignment.Center) {
                    if (cell.number != null) {
                        Text(
                            text = cell.number.toString(),
                            fontSize = 9.sp,
                            modifier = Modifier.align(Alignment.TopStart).padding(2.dp),
                            color = Color.DarkGray,
                            fontWeight = FontWeight.Bold
                        )
                    }
                    Text(
                        text = cell.userInput,
                        fontSize = 18.sp,
                        fontWeight = FontWeight.Bold,
                        color = when {
                            cell.isCorrect || isInputCorrect -> Color(0xFF2E7D32)
                            isInputWrong -> Color(0xFFD32F2F)
                            else -> Color.Black
                        }
                    )
                }
            }
        }
    }
}

File: flutter_app/lib/crossword_game_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'models/crossword.dart';
import 'models/quiz_models.dart';
import 'providers/crossword_provider.dart';
import 'widgets/mochi_background.dart';
import 'widgets/game_hud.dart';
import 'widgets/game_components.dart';
import 'providers/settings_provider.dart';

class CrosswordGameScreen extends StatelessWidget {
  const CrosswordGameScreen({super.key});

  @override
  Widget build(BuildContext context) {
    final provider = context.watch<CrosswordProvider>();
    final settings = context.watch<SettingsProvider>();

    return Scaffold(
      appBar: AppBar(
        title: Text(settings.getString("game_crosswords_title")),
        backgroundColor: Colors.transparent,
        elevation: 0,
        leading: IconButton(
          icon: const Icon(Icons.arrow_back),
          onPressed: () => _showExitDialog(context, provider),
        ),
      ),
      extendBodyBehindAppBar: true,
      body: MochiBackground(
        child: Stack(
          children: [
            SafeArea(
              child: Column(
                children: [
                  GameHUD(
                    primaryLabel: settings.getString("game_crosswords_title").toUpperCase(),
                    primaryValue: "",
                    secondaryLabel: settings.getString("game_kana_link_time_label").toUpperCase(),
                    secondaryValue: "",
                    timeSeconds: provider.gameTimeSeconds,
                  ),
                  Expanded(
                    child: InteractiveViewer(
                      boundaryMargin: const EdgeInsets.all(100),
                      minScale: 0.5,
                      maxScale: 3.0,
                      child: Center(
                        child: _buildGrid(context, provider),
                      ),
                    ),
                  ),
                  _buildClueArea(context, provider),
                  _buildKeyboard(context, provider),
                ],
              ),
            ),
            if (provider.isFinished)
              GameResultOverlay(
                isVictory: true,
                title: settings.getString("game_crossword_congrats"),
                stats: [
                  MapEntry(
                    settings.getString("game_crossword_history_item", [
                      provider.placedWords.length,
                      provider.selectedMode == CrosswordMode.kanjis ? "KANJI" : "KANA/MEANING"
                    ]),
                    _formatGameTimeHUD(provider.gameTimeSeconds),
                  ),
                ],
                onReplayClick: () => provider.startGame("n5"),
                onMenuClick: () => Navigator.pop(context),
              ),
          ],
        ),
      ),
    );
  }

  String _formatGameTimeHUD(int seconds) {
    final int m = seconds ~/ 60;
    final int s = seconds % 60;
    return m > 0 ? "${m}m ${s}s" : "${s}s";
  }

  Widget _buildGrid(BuildContext context, CrosswordProvider provider) {
    const double cellSize = 40.0;
    const int gridSize = 16;
    final theme = Theme.of(context);

    return Container(
      width: gridSize * cellSize,
      height: gridSize * cellSize,
      color: theme.colorScheme.onSurface.withOpacity(0.1),
      child: Stack(
        children: provider.cells.map((cell) {
          if (cell.isBlack) return const SizedBox.shrink();

          final isSelected = provider.selectedRow == cell.r && provider.selectedCol == cell.c;
          final isInputCorrect = cell.userInput.isNotEmpty && cell.userInput == cell.solution;
          final isInputWrong = cell.userInput.isNotEmpty && cell.userInput != cell.solution;

          final Color backgroundColor = isInputCorrect || cell.isCorrect
              ? const Color(0xFFC8E6C9)
              : (isInputWrong
                  ? const Color(0xFFFFCDD2)
                  : (isSelected ? theme.colorScheme.primaryContainer : Colors.white));

          return Positioned(
            left: cell.c * cellSize,
            top: cell.r * cellSize,
            width: cellSize,
            height: cellSize,
            child: GestureDetector(
              onTap: () => provider.onCellSelected(cell.r, cell.c),
              child: Container(
                margin: const EdgeInsets.all(0.5),
                decoration: BoxDecoration(
                  color: backgroundColor,
                  border: Border.all(
                    color: isSelected ? theme.colorScheme.primary : Colors.grey.withOpacity(0.5),
                    width: isSelected ? 1.5 : 0.5,
                  ),
                  borderRadius: BorderRadius.circular(2),
                ),
                child: Stack(
                  children: [
                    if (cell.number != null)
                      Positioned(
                        top: 2,
                        left: 2,
                        child: Text(
                          cell.number.toString(),
                          style: const TextStyle(fontSize: 8, fontWeight: FontWeight.bold, color: Colors.black54),
                        ),
                      ),
                    Center(
                      child: Text(
                        cell.userInput,
                        style: TextStyle(
                          fontSize: 18,
                          fontWeight: FontWeight.bold,
                          color: isInputCorrect || cell.isCorrect
                              ? const Color(0xFF2E7D32)
                              : (isInputWrong ? const Color(0xFFD32F2F) : Colors.black),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
          );
        }).toList(),
      ),
    );
  }

  Widget _buildClueArea(BuildContext context, CrosswordProvider provider) {
    if (provider.selectedRow == null) return const SizedBox.shrink();
    final theme = Theme.of(context);

    // Trouver le mot actif pour afficher l'indice
    final words = provider.placedWords.where((w) {
      if (w.isHorizontal) {
        return w.row == provider.selectedRow && provider.selectedCol! >= w.col && provider.selectedCol! < w.col + w.word.length;
      } else {
        return w.col == provider.selectedCol && provider.selectedRow! >= w.row && provider.selectedRow! < w.row + w.word.length;
      }
    }).toList();

    if (words.isEmpty) return const SizedBox.shrink();
    final activeWord = provider.isVerticalInput
        ? (words.firstWhere((w) => !w.isHorizontal, orElse: () => words.first))
        : (words.firstWhere((w) => w.isHorizontal, orElse: () => words.first));

    final String clueText = provider.selectedMode == CrosswordMode.kanjis
        ? "${activeWord.phonetics} (${activeWord.meaning})"
        : "${activeWord.kanji} (${activeWord.meaning})";

    return Container(
      width: double.infinity,
      padding: const EdgeInsets.all(12),
      margin: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
      decoration: BoxDecoration(
        color: theme.colorScheme.secondaryContainer,
        borderRadius: BorderRadius.circular(12),
        boxShadow: [BoxShadow(color: Colors.black.withOpacity(0.05), blurRadius: 4)],
      ),
      child: Text(
        clueText,
        textAlign: TextAlign.center,
        style: TextStyle(
          fontSize: provider.selectedMode == CrosswordMode.kanjis ? 22 : 18,
          fontWeight: FontWeight.bold,
          color: provider.selectedMode == CrosswordMode.kanjis ? theme.colorScheme.primary : theme.colorScheme.onSecondaryContainer,
        ),
      ),
    );
  }

  Widget _buildKeyboard(BuildContext context, CrosswordProvider provider) {
    if (provider.selectedRow == null || provider.isFinished) return const SizedBox.shrink();
    final theme = Theme.of(context);

    return Container(
      padding: const EdgeInsets.all(8),
      color: theme.colorScheme.surface.withOpacity(0.95),
      child: Row(
        children: [
          Expanded(
            child: Wrap(
              spacing: 4,
              runSpacing: 4,
              alignment: WrapAlignment.center,
              children: provider.keyboardKeys.map((key) {
                return SizedBox(
                  width: 50,
                  height: 50,
                  child: ElevatedButton(
                    onPressed: () => provider.onKeyTyped(key),
                    style: ElevatedButton.styleFrom(
                      padding: EdgeInsets.zero,
                      backgroundColor: theme.colorScheme.primaryContainer,
                      foregroundColor: theme.colorScheme.onPrimaryContainer,
                      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
                    ),
                    child: Text(key, style: const TextStyle(fontSize: 20, fontWeight: FontWeight.w500)),
                  ),
                );
              }).toList(),
            ),
          ),
          Container(
            width: 60,
            height: 110,
            padding: const EdgeInsets.only(left: 8),
            child: ElevatedButton(
              onPressed: () => provider.onKeyTyped(""), // Delete
              style: ElevatedButton.styleFrom(
                backgroundColor: theme.colorScheme.errorContainer,
                foregroundColor: theme.colorScheme.onErrorContainer,
                shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
              ),
              child: const Text("⌫", style: TextStyle(fontSize: 24)),
            ),
          ),
        ],
      ),
    );
  }

  void _showExitDialog(BuildContext context, CrosswordProvider provider) {
    provider.pauseGame();
    showDialog(
      context: context,
      builder: (context) => ExitConfirmationDialog(
        onConfirm: () {
          Navigator.pop(context); // close dialog
          provider.abandonGame();
          Navigator.pop(context); // exit screen
        },
        onDismiss: () {
          provider.resumeGame();
          Navigator.pop(context); // close dialog
        },
        onPause: provider.pauseGame,
        onResume: provider.resumeGame,
        onSaveAndExit: () {
          Navigator.pop(context); // close dialog
          provider.saveAndExit();
          Navigator.pop(context); // exit screen
        },
      ),
    );
  }
}
